staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_controller.go

```diff
@@ -0,0 +1,150 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apiserver
+
+import (
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/go-openapi/spec"
+	"github.com/golang/glog"
+
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/kube-aggregator/pkg/apis/apiregistration"
+)
+
+const (
+	successfulUpdateDelay   = time.Minute
+	failedUpdateMaxExpDelay = time.Hour
+)
+
+// OpenAPIAggregationManager is the interface between this controller and OpenAPI Aggregator service.
+type OpenAPIAggregationManager interface {
+	AddUpdateApiService(handler http.Handler, apiService *apiregistration.APIService) error
+	UpdateApiServiceSpec(apiServiceName string, spec *spec.Swagger, etag string) error
+	RemoveApiServiceSpec(apiServiceName string) error
+	GetApiServiceInfo(apiServiceName string) (handler http.Handler, etag string, exists bool)
+}
+
+// OpenAPIAggregationController periodically check for changes in OpenAPI specs of APIServices and update/remove
+// them if necessary.
+type OpenAPIAggregationController struct {
+	openAPIAggregationManager OpenAPIAggregationManager
+	queue                     workqueue.RateLimitingInterface
+	downloader                *openAPIDownloader
+}
+
+func NewOpenAPIAggregationController(downloader *openAPIDownloader, openAPIAggregationManager OpenAPIAggregationManager) *OpenAPIAggregationController {
+	c := &OpenAPIAggregationController{
+		openAPIAggregationManager: openAPIAggregationManager,
+		queue: workqueue.NewNamedRateLimitingQueue(
+			workqueue.NewItemExponentialFailureRateLimiter(successfulUpdateDelay, failedUpdateMaxExpDelay), "APIServiceOpenAPIAggregationControllerQueue1"),
+		downloader: downloader,
+	}
+
+	return c
+}
+
+func (c *OpenAPIAggregationController) Run(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	glog.Infof("Starting OpenAPIAggregationController")
+	defer glog.Infof("Shutting down OpenAPIAggregationController")
+
+	go wait.Until(c.runWorker, time.Second, stopCh)
+
+	<-stopCh
+}
+
+func (c *OpenAPIAggregationController) runWorker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+// processNextWorkItem deals with one key off the queue.  It returns false when it's time to quit.
+func (c *OpenAPIAggregationController) processNextWorkItem() bool {
```

Comment:
This looks pretty close to a controller now.  This one should follow the normal conventions from the rest of our controllers and it will be better separated and easier for general maintainers to inspect/fix/contribute.

Reply:
See https://github.com/kubernetes/community/blob/master/contributors/devel/controllers.md for the conventional form of processNextWorkItem.  Substituting just `syncHandler` makes the error handling easy to see as "standard" and keeps your logic out of the error handling paths.

Reply:
Done in next commit.